src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java

```diff
@@ -61,20 +64,35 @@ public void run() {
    */
   private void runNodes(List<T> freeNodes) {
     List<IWorker<T>> runnables = m_factory.createWorkers(freeNodes);
+    CountDownLatch latch = null;
+    if (useLatches && runnables.size() > 1) {
+      latch = new CountDownLatch(runnables.size());
+    }
+    boolean latchUsed = false;
     for (IWorker<T> r : runnables) {
       m_activeRunnables.add(r);
+      if (latch != null && (r instanceof TestMethodWorker)) {
+        //Being conservative here and setting this only for test method workers
+        r.setLatch(latch);
+        latchUsed = true;
+      }
       ppp("Added to active runnable");
       setStatus(r, Status.RUNNING);
       ppp("Executing: " + r);
       try {
         execute(r);
-//        if (m_threadCount > 1) execute(r);
-//        else r.run();
       }
       catch(Exception ex) {
         ex.printStackTrace();
       }
     }
+    if (latch != null && latchUsed) {
+      try {
+        latch.await(10, TimeUnit.SECONDS);
```

Comment:
@juherr -Your approach would basically stall at every method for a max of 10 seconds no ?

w.r.t the iteration classification, yes, that was what I was hoping for as well, but its not like that. I still don't have enough understanding of `DynamicGraph` to be able to figure out where the problem is. Still trying to understand its logic.